### PR TITLE
Fixes display of speed and ATP consumption of flagella

### LINKED ISSUE #2406: Fix flagella draining atp when visually inactive
When having multiple flagella, with only one visually (and normally) active, for a build able to sustain it (from the editor charts), the atp still decreases.

Further tests highlighted that this could occur for flagella colinearly and (especially) diagonally opposed to the movement.
This occurs even when flagella are placed with valid editor rotation, although it seems more to be in see-saw (major drop then atp rises again).

Perhaps comes from proportional reduction of flagella? 

Perhaps related to #1500 and #1297 . 
_____
Screens from initial, replicated report

One flagella sustainability
image

Final design 
image

Replication with one opposed flagella

image

### PR DESCRIPTION
**Pull request by Kemikal1**
**Brief Description of What This PR Does**

Fixes the speed and ATP cost display.

**Related Issues**
Fixes #2406 

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author.

### CODE REVIEW

**Comment by Kemikal1** [comment]
Turns out it was harder then i bargained for.... oufff

**Comment by buckly90** [comment]
Seems to work without any visible error.

I noticed that this makes it so that backwards facing flagella display positive speed instead of negative. When I place two flagella facing opposite directions, it results in an overall decrease in speed. Am I mistaken in assuming that it calculates top speed in one direction? Whatever is the case, it is pulled off quite well so excellent job!

**Comment by Kemikal1** [comment]
Thats indeed the intended function, because a microbe cant go in opossite direction at thr same time. Well for a given microbe theres a single direction that represent the microbes top speed. Maybe we could also display that direction in a new PR.

File: src/microbe_stage/MicrobeInternalCalculations.cs [hunk 1]
```diff
@@ -1,39 +1,147 @@
...
+        var organelles = inputOrganelles.Where(o => o.Definition.HasComponentFactory<MovementComponentFactory>())
+            .ToList();
+
+        foreach (var organelle in organelles)
+        {
+            maximumMovementDirection += GetOrganelleDirection(organelle);
+        }
+
+        // After calculating the sum of all organelle directions we substract the movement components which
+        // are symetric and we chose the one who would benefit the max-speed the most.
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeInternalCalculations.cs:20]:
```suggestion
        // are symetric and we choose the one who would benefit the max-speed the most.
```

File: src/microbe_stage/MicrobeInternalCalculations.cs [hunk 2]
```diff
@@ -1,39 +1,147 @@
...
+
+    private static float MovementForce(float movementForce, float directionFactor)
+    {
+        if (directionFactor < 0)
+            return 0;
+
+        return movementForce * directionFactor;
+    }
+
+    // Symetric flagella are a corner case for speed calculations because the sum of all
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeInternalCalculations.cs:121]: This should be an xml remark

File: src/microbe_stage/ProcessSystem.cs [hunk 3]
```diff
@@ -49,14 +49,17 @@ public ProcessSystem(Node worldRoot)
...
+        return ComputeEnergyBalance(organelles, biome, membrane, maximumMovementDirection);
     }
 
     /// <summary>
     ///   Computes the energy balance for the given organelles in biome
     /// </summary>
-    public static EnergyBalanceInfo ComputeEnergyBalance(IEnumerable<OrganelleDefinition> organelles,
-        BiomeConditions biome, MembraneType membrane)
+    public static EnergyBalanceInfo ComputeEnergyBalance(IEnumerable<OrganelleTemplate> organelles,
+        BiomeConditions biome, MembraneType membrane, Vector3 maximumMovementDirection)
```
Comment (hhyyrylainen) [inline src/microbe_stage/ProcessSystem.cs:62]: `maximumMovementDirection` should be renamed here to something that explains it is used to filter out movementOrganelles that aren't active if they don't allow moving in that direction. Maybe something like `excludeMovementIfCannotPushInThisDirection` though that's a bit wordy

File: src/microbe_stage/MicrobeInternalCalculations.cs [hunk 4]
```diff
@@ -1,39 +1,147 @@
...
+    // directions is kinda of broken in their case, so we have to choose which one of the symmetric flagella
+    // we must discard from the direction calculation
+    // Here we only discared if the flagella we input is the "bad" one
+    private static Vector3 ChooseFromSymetricFlagella(IEnumerable<OrganelleTemplate> inputOrganelles,
+        OrganelleTemplate testedOrganelle, Vector3 maximumMovementDirection)
+    {
+        var organelles = inputOrganelles.ToList();
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeInternalCalculations.cs:128]: This ToList call is completely unnecessary

File: src/microbe_stage/MicrobeInternalCalculations.cs [hunk 5]
```diff
@@ -1,39 +1,147 @@
 ﻿using System.Collections.Generic;
+using System.Linq;
 using Godot;
 
 public static class MicrobeInternalCalculations
 {
+    public static Vector3 MaximumSpeedDirection(IEnumerable<OrganelleTemplate> inputOrganelles)
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeInternalCalculations.cs:7]: Why is this called `inputOrganelles`I think the parameter should just be called `organelles` so that it doesn't confuse the future code writer as to why this takes in "inputOrganelles" and not just "organelles"?
You can rename the local variable to movementOrganelles.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
I guess all the parts are necessary, as it's realistic to only calculate the maximum movement force providing flagella energy consumption.
I have some code clarity related comments, I was not able to understand all of the code, but I can let that slide, I didn't really understand the previous speed calculation math either...

**Review by RacerBG — APPROVED** [review]

File: src/microbe_stage/MicrobeInternalCalculations.cs (hunk 4 above)
Comment (Kemikal1) [reply src/microbe_stage/MicrobeInternalCalculations.cs:128]: jetbrains gives me an error for multiple enumerations if i iterate multiple times through it.

File: src/microbe_stage/ProcessSystem.cs (hunk 3 above)
Comment (Kemikal1) [reply src/microbe_stage/ProcessSystem.cs:62]: why should i rename  a variable like that? thats not what the variable represents. Armed with some basic vector math and seeing its use in calcualtions you should know in what way and why the variable is used.

File: src/microbe_stage/MicrobeInternalCalculations.cs (hunk 4 above)
Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeInternalCalculations.cs:128]: But you do not? I don't see the organelles variable being used anywhere else in this method than on the next line.

File: src/microbe_stage/ProcessSystem.cs (hunk 3 above)
Comment (hhyyrylainen) [reply src/microbe_stage/ProcessSystem.cs:62]: It doesn't fit the design of ComputeEnergyBalance. It should not have to know that maximum movement direction is the only direction flagella are ever used in. This is a lower level building block that should be made with the assumption that whoever is looking at it doesn't know anything about maximum movement direction, just that if you want this lower level method can exclude any movement organelles that can't move the cell in a certain direction from the computation.

Comment (Kemikal1) [reply src/microbe_stage/ProcessSystem.cs:62]: Totally disagree, when computing the maximum energy consumption of a microbe the direction in which the speed and, thus, the amount of flagella activation is the greatest is pretty relevant since both are tied together.

File: src/microbe_stage/MicrobeInternalCalculations.cs (hunk 4 above)
Comment (Kemikal1) [reply src/microbe_stage/MicrobeInternalCalculations.cs]: yea, i didint see that there

File: src/microbe_stage/MicrobeInternalCalculations.cs [hunk 6]
```diff
@@ -118,15 +119,17 @@ private static float MovementForce(float movementForce, float directionFactor)
...
+    ///  directions is kinda of broken in their case, so we have to choose which one of the symmetric flagella
+    ///  we must discard from the direction calculation
+    ///  Here we only discard if the flagella we input is the "bad" one
+    /// </summary>
+    private static Vector3 ChooseFromSymetricFlagella(IEnumerable<OrganelleTemplate> organelles,
         OrganelleTemplate testedOrganelle, Vector3 maximumMovementDirection)
     {
-        var organelles = inputOrganelles.ToList();
-        foreach (var organelle in organelles)
+        var movementOrganelles = organelles;
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeInternalCalculations.cs:131]: Why have this separate variable assignment here?

File: src/microbe_stage/ProcessSystem.cs (hunk 3 above)
Comment (hhyyrylainen) [reply src/microbe_stage/ProcessSystem.cs:62]: I would be more willing to accept that, if this method wasn't public. This is a public method and now it has this new confusing parameter that doesn't seem to be related to this method's purpose at all. I might accept this naming if you add xml parameter comments.

Comment (Kemikal1) [reply src/microbe_stage/ProcessSystem.cs:62]: For all the parameters?

**Comment by hhyyrylainen** [comment]
I had this TODO note:
After merging this need to open an issue for auto-evo to consider the maximum speed in a direction so it can't cheat by placing sideways flagella that don't cost energy but in the simple calculations they give movement speed. Probably should cache the maximum speed direction and speed value for efficiency

File: src/microbe_stage/ProcessSystem.cs (hunk 3 above)
Comment (hhyyrylainen) [reply src/microbe_stage/ProcessSystem.cs:62]: I'll do this myself.

**Review by hhyyrylainen — APPROVED** [review]
Did the final fixes I wanted. I still don't fully understand why I got the playtesting results I got, but I'll merge this now anyway.
